Branch streamice updates

dngoldberg opened the pull request:
## What changes does this PR introduce?
minor changes to pkg/streamice:
a) allow option to specify a "regularised coulomb" sliding law (as described in change to docs)
b) allow surface mass balance/accumulation to be specified via file
c) allow a max. num of lin solves to be carried out using PETSc (when PETSc enabled) for a time step
d) fix a backward compatibility issue found to occur when using PETSc
e) make use of an incomplete Cholesky factorization when using the PETSc BlockJacobi Preconditioner
f) introduce a parameter to control the fill level in (e)

## What is the current behaviour? 
a) this does not exist
b) smb must be uniform
c) either all petsc or all non-petsc lin solves
d) uses petsc initialisation variable PETSC_DEFAULT_DOUBLE_PRECISION which has been deprecated in some petsc versions
e) uses incomplete ILU precond (PCILU)
f) no such runtime parameter

## What is the new behaviour 
a) option to specify a "regularised coulomb" sliding law (as described in change to docs)
b) surface mass balance/accumulation can be specified via file
c) first N solves is carried out with PETSc if option is used
d) uses petsc initialisation variable PETSC_DEFAULT_REAL which has been deprecated in some petsc versions
e) uses incomplete Chol precond (PCICC) to preserve self adjoint property
f) run time parameter streamice_petsc_pcfactorlevels introduced


## Does this PR introduce a breaking change? 
no


## Other information:


## Suggested addition to `tag-index`
o pkg/streamice:
  - add option to specify "regularised coulomb" sliding law to better reflect
    dependence on subglacial water pressure close to the grounding line
  - above requires CPP option (due to new arrays) STREAMICE_COULOMB_SLIDING
    and r/t param streamice_allow_reg_coulomb
  - add diagnostic for streamice melt rate (BDOT_SI )
  - add parameter STREAMICEAdotFile to specify surface mass balance
  - add parameter streamice_maxnliter_Petsc to specify maximum no. of 
    linear solves using PETSc in a single time step (not incl. in doc
    since PETSc is not part of standard build
  - update references to PETSC_DEFAULT_DOUBLE_PRECISION to 
    PETSC_DEFAULT_REAL in streamice_cg_solve_petsc.F; former seems to be
    deprecated in some versions of PETSc while latter is backward compatible
  - for BJACOBI (block jacobi) PETSc preconditioner, use PCICC (incomplete
    cholesky) rather than PCILU (incomplete LU)
  - for BJACOBI add parameter streamice_petsc_pcfactorlevels to control
    level of factor fill-in

jm-c commented:
@dngoldberg I took a look to the code (and made little cleaning) but waiting to get a proper PETC
installed to get things tested. 
@jrscott If you could review the addition to the docs, it would be helpful.

dngoldberg commented:
hi @jm-c could you clarify, do you need to install PETSc before being able to approve?

jm-c commented:
Hi Dan,

I will see how long it takes, hopefully it will be quick.
If it takes too long I will approve it without testing, just relying on
you having tested it.

Jean-Michel

On Mon, Feb 24, 2020 at 03:08:35PM -0800, dngoldberg wrote:
> hi @jm-c could you clarify, do you need to install PETSc before being able to approve?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/328#issuecomment-590595076

jm-c commented:
@dngoldberg I forgot this: It would be nice to add one line of description in STREAMICE.h
for the new parameters that you added in this PR. It's a good practice and it's helpful.

dngoldberg commented:
@jm-c  it should work with any version above 3.6. here is my configuration --
Configure Options: --configModules=PETSc.Configure --optionsModule=PETSc.compilerOptions --configModules=PETSc.Configure --optionsModule=PETSc.compilerOptions --configModules=PETSc.Configure --optionsModule=PETSc.compilerOptions --configModules=PETSc.Configure --optionsModule=PETSc.compilerOptions --CFLAGS="-fPIC -march=native" --CXXFLAGS="-fPIC -march=native" --FFLAGS="-fPIC -march=native" --with-debugging=0 --with-hypre=1 --download-hypre=1 --with-umfpack=1 --download-umfpack=1 --with-spooles=1 --download-spooles=1 --with-mpi=1 --with-mpi-interfaces=1 --with-fortran-interfaces=1 -with-shared-libraries=1 --with-mumps=1 --download-mumps=1 --with-scalapack=1 --download-scalapack=1 --with-blacs=1 --download-blacs=1 --with-parmetis=1 --download-parmetis=1 --with-metis=1 --download-metis=1 --download-f-blas-lapack=1

Please note, that if MITgcm is to compile and run with PETSc, linking and include paths need to be added to the makefile for the PETSc binaries e.g. through an options file. I am aware that this should be done more robustly, but (a) these binaries and header files can appear in different locations depending on the installation, (b) for different versions of PETSc the fortran include files can appear in slightly different locations in the PETSc source tree. a proper solution, i believe, would be a way to pass in the location of the PETSc installation as a command line argument to genmake2, and furthermore have genmake2 detect the version. However I would not even know where to begin in terms of modifying genmake2 as i do not understand its workings at all and that is why i have kept mention of PETSc out of the docs. At one point we had discussed with Chris using PETSc to solve cg3d, but i don't know if this has progressed. Please let me know what you would like to do. 

Understood regarding STREAMICE.h -- should i modify?

dngoldberg commented:
hi @jm-c can i ask you to hold off on reviewing / approving PR? while nothing is interfering w verification there may be some interactions between parameters i did not foresee. Moreover as you say, the STREAMICE.h new variables should have explanation, so i will add those as well.

i want to avoid having to make multiple PRs because i did not check things carefully enough the first time..

dngoldberg commented:
@jm-c have had a chance to do more testing and have addressed the issues i saw previously. 

I have also included an additional change regarding petsc functionality, as it is minor and there are extensive changes in this P/R i have also decided to include it in this P/R for your review. the main modifications are to streamice_cg_solve_petsc.F as well as an additional parameter streamice_petsc_pcfactorlevels added to STREAMICE.h and streamice_readparms.F. this is detailed in tag-index. I hope this is OK -- i figured this would be easier than submitting an additional PR.

As requested i have also added comments describing all parameters in STREAMICE.h. I have identified some defunct parameters but need to make sure that they can be deleted, i will investigate further for a future P/R.

jm-c commented:
A bit lagging behind: It seem Dan finished to make the modifs he wanted to add so shoud go fro reviewing to move this PR towards merging.
@jrscott can you take a look at the doc changes ?
And I will check latest file changes soon (it does not seem like I will get a working PETC installed soon so likely to pass this step).

jrscott commented:
@jm-c It does not look like there have been any changes to doc since I reviewed a month ago (?)

dngoldberg commented:
@jrscott you reviewed for this P/R yes? no there have not been changes.

jrscott commented:
@jm-c I will try to get an update on the svante PETSc install today. Eventually I will get this done...

jrscott commented:
@dngoldberg yes I went through your doc changes, made some very minor edits, see my changes above.

dngoldberg commented:
@jrscott  thanks very much am happy with this. -dan

jm-c commented:
I was not able to check with PETSC but the rest looks good. 
Will merge later today or tomorrow.
